// console-feed-modern/lib/Component/Message.js
import * as React from 'react';
import { ThemeProvider } from 'emotion-theming';
import { Message, Icon, Content } from './elements';
import Formatted from './message-parsers/Formatted';
import ObjectTree from './message-parsers/Object';
import ErrorPanel from './message-parsers/Error';
class ConsoleMessage extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.theme = (theme) => ({
            ...theme,
            method: this.props.log.method
        });
    }
    render() {
        const { log } = this.props;
        return (React.createElement(ThemeProvider, { theme: this.theme },
            React.createElement(Message, { "data-method": log.method },
                React.createElement(Icon, null),
                React.createElement(Content, null, this.getNode()))));
    }
    getNode() {
        const { log } = this.props;
        // Error handling
        const error = this.typeCheck(log);
        if (error)
            return error;
        // Chrome formatting
        if (log.data.length > 0 &&
            typeof log.data[0] === 'string' &&
            log.data[0].indexOf('%') > -1) {
            return React.createElement(Formatted, { data: log.data });
        }
        // Error panel
        if (log.data.every(message => typeof message === 'string') &&
            log.method === 'error') {
            return React.createElement(ErrorPanel, { log: log });
        }
        // Normal inspector
        const quoted = typeof log.data[0] !== 'string';
        return React.createElement(ObjectTree, { log: log, quoted: quoted });
    }
    typeCheck(log) {
        if (!log) {
            return (React.createElement(Formatted, { data: [
                    `%c[console-feed] %cFailed to parse message! %clog was typeof ${typeof log}, but it should've been a log object`,
                    'color: red',
                    'color: orange',
                    'color: cyan'
                ] }));
        }
        else if (!(log.data instanceof Array)) {
            return (React.createElement(Formatted, { data: [
                    '%c[console-feed] %cFailed to parse message! %clog.data was not an array!',
                    'color: red',
                    'color: orange',
                    'color: cyan'
                ] }));
        }
        return false;
    }
}
export default ConsoleMessage;
